### DISCUSSION TITLE
refetch vs invalidating query

### DISCUSSION THREAD

**Discussion by ytnytn1** [TanStack/query#2468] [Q&A] [answered]
Hello! 
In the [documentation ](https://react-query.tanstack.com/guides/invalidations-from-mutations) how to invalidate queries after mutations I can see that it is suggested to use query invalidation. I'm curious what if instead of invalidation use refetching ?  Won't give the same effect?  

Thanks in advance!

**Comment by TkDodo** [comment, marked as answer]
>  I'm curious what if instead of invalidation use refetching ? Won't give the same effect?

invalidation is a more "smart" refetching. Refetching will always refetch, even if there are no observers for the query. invalidation will just mark them as stale so that they refetch the next time an observer mounts. For queries with active observers, there is no difference as far as I'm aware.

**Reply by ytnytn1** [reply]
Thanks a lot! Now it is clear to me.

**Reply by aegiz** [reply]
I had the same exact question as @ytnytn1 and came to the same conclusion as mentioned by @TkDodo 
Can I suggest to add a quick line in the documentation to make it more clear what the differences are between `refetch` and `invalidate`. Thanks

**Reply by TkDodo** [reply]
@aegiz please open a PR to the docs with a proposed change if you feel we can make the docs better 👍

**Reply by aegiz** [reply]
Yes, while doing so I actually found something interesting: [here](https://react-query.tanstack.com/reference/QueryClient#queryclientinvalidatequeries) it says:
"If you do not want active queries to refetch, and simply be marked as invalid, you can use the refetchActive: false option.
If you want inactive queries to refetch as well, use the refetchInactive: true option"

So what's the difference between: 
```
await queryClient.invalidateQueries('company', {
      refetchActive: true, // default
      refetchInactive: true, // force inactive queries to refetch
  });
```
and 
```
await companyQuery.refetch();
```
Is there one that you would encourage over the other? @TkDodo 
Thanks

**Reply by TkDodo** [reply]
you usually don't have access to `.refetch()` when you want to `invalidate`, as refetch is returned from the `useQuery` hook. Also, `invalidateQueries` has support for fuzzy query key matching

**Reply by aegiz** [reply]
Make perfect sense, thank you!

**Reply by muditdev** [reply]
All make sense, i feel invalidate is better as it is much easier to invalidate by query key. no need to look for refetch fn if working with nested components.

**Reply by nvmnghia** [reply]
Is there a difference between `refetch` and `prefetch` then? Both seem to fetch right away.

**Reply by TkDodo** [reply]
`prefetchQueries` is the same as  `fetchQueries` under the hood, except that it doesn't return anything or throw errors, so you can just `await` it without handling errors, as with pre-fetching, it's usually fine if it doesn't work.

Also, it respects the `staleTime` that you pass to it (or that has been globally set), so if data is fresh, it will be resolved immediately.

`refetch` is just a hardcore "fetch now, no questions asked" (at leat the one returned from useQuery). `refetchQueries` still respects `enabled:false` for example.

**Reply by dgd03146** [reply]
Then It seems to invalidate queries is always a better option. Is there any time to use refetch instead of using invalidate queries?

**Reply by TkDodo** [reply]
I don't use refetch :)

**Reply by markivancho** [reply]
Also, be aware that with `refetch` your query won't care about the `enabled` option. It will trigger the `queryFn` regardless of it. Invalidate, on the other hand, will work just like a charm

**Reply by TkDodo** [reply]
That's true for `refetch` returned from `useQuery`, but not for `queryClient.refetchQueries` (this respects enabled) :)

**Reply by TamirCode** [reply]
invalidate will only do the refetch when the user sees the component with the data. But if you want the data to be fetched and prepared before the user accesses the data then it would be better for smoother UX to fetch in the background beforehand by using refetch()

**Reply by markivancho** [reply]
@TamirCode I believe there is a better alternative - `await queryClient.prefetchQuery({ queryKey, queryFn })` for your use case

**Reply by paulosborne** [reply]
@TkDodo is there a reason why `useQuery` doesn't return an `invalidate` function alongside `refetch`? it feels a bit janky to have to import `useQueryClient` and then call `invalidateQueries` or `refetchQueries`

**Reply by TkDodo** [reply]
`refetch` is just returned because it's needed to bypass `enabled: false` for an imperative approach (that I personally don't like).

Usually, invalidation and showing the actual query doesn't happen in the same component. Invalidation often happens as a result of a mutation, which means you'll have a custom hook somewhere that wraps `useMutation`. That hook doesn't have access to the return values of `useQuery`, so it's not helping here. Cases where you have both in the same component are rare and adding 1 line of code to those isn't really a problem

**Reply by jgarplind** [reply]
I think I have found a difference that should probably be considered under [certain circumstances](https://github.com/TanStack/query/discussions/6953#discussioncomment-8983467): `invalidateQueries` will deduplicate in-flight requests preferring the first sent request (see https://github.com/TanStack/query/discussions/608) whereas `refetch` (as well as `refetchQueries`) will cancel in-flight requests i.e. preferring the last sent request (see bottom of https://tanstack.com/query/latest/docs/framework/react/reference/useQuery).

Typically, this is not a concern, but in rare scenarios this can matter, e.g.:
1. GET-request for key X already in-flight, with server-side logic finished before no. 2
2. mutation finishes, and requests another cache invalidation for key X (using the common pattern `onSuccess: invalidate X`)
3. Since the GET-request is still in-flight, no. 2 never causes a new request
4. Now the GET-request returns, but it contains stale data, since any server-side logic was run _before_ no. 2

While this timing seems unlikely, it has happened enough to be a nuissance but too little to be reproducable on demand in our chatty CQRS-oriented architecture.

Now, I am hoping that a few strategic replacements of `invaliateQueries` in favour of `refetch`/`refetchQueries` will reduce the risk for stale state to practically impossible.

**Reply by TkDodo** [reply]
> invalidateQueries will deduplicate in-flight requests preferring the first sent request

no it doesn't. you're referencing a discussion from 2020, and this behaviour was changed in v4:

- https://tanstack.com/query/v5/docs/framework/react/guides/migrating-to-react-query-4#consistent-behavior-for-cancelrefetch

**Reply by toytonic** [reply]
Sorry for chipping in here @TkDodo, but this is the discussion that I found closest to answer my question. 
Is there any other way to bypass `enbaled` than `refetch`? I have the exact use case where I have a disabled query for a search/filter result only executing on user click, but mutations in nested components that should trigger a refetch for updating the result list.

**Reply by TkDodo** [reply]
> Is there any other way to bypass enbaled than refetch? 

no.

> I have the exact use case where I have a disabled query for a search/filter result only executing on user click

can you show a reproduction please? I don’t get that use-case. `enabled: !!myCondition` usually means that the query _cannot_ be executed until `myCondition` is fulfilled. Why would you need to run it anyways? If you have `enabled: false` (like, the hardcoded value `false`) for search/filtering onClick, you’re doing it wrong.

**Reply by toytonic** [reply]
Thanks for clarifying @TkDodo.
I might very well have chosen the wrong path or went for composing my fetch/mutation logic the wrong way. 

The use-case is that I want to fetch some list data on user click. AFAIK I have to set `enabled: false` and use `refetch` to achieve this. After the result is fetched, I can mutate/delete individual items of that result list. In order to update to the result list after successful mutation/deletion I have to update the list. Of course I can (optimistically) manipulate the cache, but best practice according to examples is to refetch anyways, right?

I placed the mutation/deletion logic into custom hooks, which as of now do not have access to the `refetch` function of the result list query.  Which they would not need to if I could make the result list query refetch through invalidation.

So hopefully that brings some clarity on my specific use-case. Happy to receive some feedback how to improve things on my end in general.

**Reply by TkDodo** [reply]
> AFAIK I have to set enabled: false and use refetch to achieve this

you usually want lazy queries: https://tanstack.com/query/latest/docs/framework/react/guides/disabling-queries#lazy-queries
